lib/list_versions.go

```diff
@@ -1,142 +1,166 @@
-func GetTFLatest(mirrorURL string) (string, error) {
+// getReleases : subfunc for GetTFReleases, used in a loop to get all terraform releases given the hashicorp url
+func getReleases(mirrorURL string, queryParams map[string]string) ([]*Release, error) {
+	var releases []*Release
+	resp, errURL := httpGet(mirrorURL, queryParams)
+	if errURL != nil {
+		return nil, fmt.Errorf("[Error] : Getting url: %v", errURL)
+	}
+	defer resp.Body.Close()
 
-	result, error := GetTFURLBody(mirrorURL)
-	if error != nil {
-		return "", error
-	}
-	// Getting versions from body; should return match /X.X.X/ where X is a number
-	semver := `\/(\d+\.\d+\.\d+)\/?"`
-	r, _ := regexp.Compile(semver)
-	for i := range result {
-		if r.MatchString(result[i]) {
-			str := r.FindString(result[i])
-			trimstr := strings.Trim(str, "/\"") //remove '/' or '"' from /X.X.X/" or /X.X.X"
-			return trimstr, nil
-		}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("[Error] : Retrieving contents from url: %s", mirrorURL)
 	}
 
-	return "", nil
-}
+	body := new(bytes.Buffer)
+	if _, err := io.Copy(body, resp.Body); err != nil {
+		return nil, err
+	}
 
-//GetTFLatestImplicit :  Get the latest implicit terraform version given the hashicorp url
-func GetTFLatestImplicit(mirrorURL string, preRelease bool, version string) (string, error) {
+	if err := json.Unmarshal(body.Bytes(), &releases); err != nil {
+		return nil, fmt.Errorf("%s: %s", err, body.String())
+	}
+	return releases, nil
+}
 
-	if preRelease == true {
-		//TODO: use GetTFList() instead of GetTFURLBody
-		versions, error := GetTFURLBody(mirrorURL)
-		if error != nil {
-			return "", error
-		}
-		// Getting versions from body; should return match /X.X.X-@/ where X is a number,@ is a word character between a-z or A-Z
-		semver := fmt.Sprintf(`\/(%s{1}\.\d+\-[a-zA-z]+\d*)\/?"`, version)
-		r, err := regexp.Compile(semver)
-		if err != nil {
-			return "", err
-		}
-		for i := range versions {
-			if r.MatchString(versions[i]) {
-				str := r.FindString(versions[i])
-				trimstr := strings.Trim(str, "/\"") //remove '/' or '"' from /X.X.X/" or /X.X.X"
-				return trimstr, nil
-			}
-		}
-	} else if preRelease == false {
-		listAll := false
-		tflist, _ := GetTFList(mirrorURL, listAll) //get list of versions
-		version = fmt.Sprintf("~> %v", version)
-		semv, err := SemVerParser(&version, tflist)
+//GetTFReleases :  Get all terraform releases given the hashicorp url
+func GetTFReleases(mirrorURL string, preRelease bool) ([]*Release, error) {
+	limit := 20
+	queryParams := map[string]string{"limit": strconv.Itoa(limit)}
+	releaseSet, err := getReleases(mirrorURL, queryParams)
+	if err != nil {
+		return nil, err
+	}
+	var releases []*Release
+	releases = append(releases, releaseSet...)
+	for len(releaseSet) == limit {
+		queryParams["after"] = releaseSet[len(releaseSet)-1].TimestampCreated.String()
+		releaseSet, err = getReleases(mirrorURL, queryParams)
```

Comment:
This is only going to query up to 2 pages of results, right?
Is there an implicit guarantee that this will be sufficient? Or should this be refactored to loop until the results are exhausted?

Reply:
This should already loop over all pages, are you seeing different results?
It should continue looping while the size of results is equal to the specified `limit`, after which we know we've reached the last page.

Reply:
nope, just haven't had coffee and hopelessly misread the `for` as `if` 🤦‍♂️

Reply:
No worries mate!